pkg/compose/hook.go

```
@@ -0,0 +1,122 @@
+			ID:        container.ID,
+			Service:   service.Name,
+			Line:      line,
+		})
+	})
+	defer wOut.Close() //nolint:errcheck
+
+	detached := listener == nil
+	exec, err := s.apiClient().ContainerExecCreate(ctx, container.ID, containerType.ExecOptions{
+		User:         hook.User,
+		Privileged:   hook.Privileged,
+		Env:          ToMobyEnv(hook.Environment),
+		WorkingDir:   hook.WorkingDir,
+		Cmd:          hook.Command,
+		Detach:       detached,
+		AttachStdout: !detached,
+		AttachStderr: !detached,
+	})
+	if err != nil {
+		return err
+	}
+
+	if detached {
+		return s.runWaitExec(ctx, exec, service, listener)
+	}
+
+	height, width := s.stdout().GetTtySize()
+	consoleSize := &[2]uint{height, width}
+	attach, err := s.apiClient().ContainerExecAttach(ctx, exec.ID, containerType.ExecAttachOptions{
+		Tty:         service.Tty,
+		ConsoleSize: consoleSize,
+	})
+	if err != nil {
+		return err
+	}
+	defer attach.Close()
+
+	if service.Tty {
+		_, err = io.Copy(wOut, attach.Reader)
+	} else {
+		_, err = stdcopy.StdCopy(wOut, wOut, attach.Reader)
+	}
+	if err != nil {
+		return err
+	}
+
+	inspected, err := s.apiClient().ContainerExecInspect(ctx, exec.ID)
+	if err != nil {
+		return err
+	}
+	if inspected.ExitCode != 0 {
+		return fmt.Errorf("%s hook exited with status %d", service.Name, inspected.ExitCode)
+	}
+	return nil
+}
+
+func (s composeService) runWaitExec(ctx context.Context, exec moby.IDResponse, service types.ServiceConfig, listener api.ContainerEventListener) error {
+	err := s.apiClient().ContainerExecStart(ctx, exec.ID, containerType.ExecStartOptions{
+		Detach: listener == nil,
+		Tty:    service.Tty,
+	})
+	if err != nil {
+		return nil
+	}
+
+	// We miss a ContainerExecWait API
+	tick := time.NewTicker(100 * time.Millisecond)
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-tick.C:
+			inspect, err := s.apiClient().ContainerExecInspect(ctx, exec.ID)
+			if err != nil {
+				return nil
```

Might have missed something, but shouldn't we return the error in this case?